zilliqa/src/consensus.rs

```diff
@@ -1213,87 +1521,118 @@ impl Consensus {
             .timestamp()
             .elapsed()
             .unwrap_or_else(|err| err.duration());
-        if difference > self.config.allowed_timestamp_skew {
+        if difference > self.config.allowed_timestamp_skew && parent.view() > 0 {
+            warn!(
+                "timestamp difference greater than allowed skew: {difference:?}. Blocks {0:?} and {1:?}", block.view(), parent.view(),
+            );
+        }
+
+        // Blocks must be in sequential order
+        if block.header.number != parent.header.number + 1 {
             return Err(anyhow!(
-                "timestamp difference greater than allowed skew: {difference:?}"
+                "block number is not sequential: {} != {} + 1",
+                block.header.number,
+                parent.header.number
             ));
         }
 
         if !self.block_extends_from(block, &finalized_block)? {
-            return Err(anyhow!("invalid block"));
+            warn!(
+                "invalid block {:?}, does not extend finalized block {:?} our head is {:?}",
+                block,
+                finalized_block,
+                self.head_block()
+            );
+
+            return Err(anyhow!(
+                "invalid block, does not extend from finalized block"
+            ));
         }
         Ok(())
     }
 
     // Checks for the validity of a block and adds it to our block store if valid.
-    pub fn receive_block(&mut self, block: Block) -> Result<()> {
+    // Returns true when the block is valid and newly seen and false otherwise.
+    pub fn receive_block(&mut self, block: Block) -> Result<bool> {
         if self.block_store.contains_block(block.hash())? {
-            return Ok(());
+            trace!("recieved block already seen: {}", block.hash());
+            return Ok(false);
         }
 
         match self.check_block(&block) {
             Ok(()) => {
+                trace!("updating high QC and view, blocks seems good!");
                 self.update_high_qc_and_view(block.agg.is_some(), block.qc.clone())?;
                 self.add_block(block)?;
             }
             Err(e) => {
-                // TODO: Downcasting is a bit ugly here - We should probably have an error enum instead.
-                if let Some(e) = e.downcast_ref::<MissingBlockError>() {
-                    // We don't call `update_high_qc_and_view` here because the block might be a fork of the finalized chain
-                    self.add_block(block)?;
-                    match e.0 {
-                        BlockRef::Hash(hash) => self.block_store.request_block(hash),
-                        BlockRef::View(view) => self.block_store.request_block_by_view(view),
-                    }?
-                } else {
-                    return Err(e);
-                }
+                //warn!("Error when receiving block {}", e);
+                warn!(?e, "invalid block received during sync!");
+                //return Err(e);
```

Comment:
(dev comments?)

Reply:
ah yes, good catch. 

As a note: this is one of the bugs that was fixed - when the network is restarted from scratch, it is possible for a block to be sent to you which fails the block check. Rather than making an err, just return the block was bad.

Reply:
Nice